ext/storage_sources/s3_store/s3_storage_source.cpp

```diff
@@ -55,11 +62,139 @@ const uint64_t partSize = 8 * 1024 * 1024; /* 8 MB. */
 /* Setting SDK options. */
 Aws::SDKOptions options;
 
+static int S3GetDirectory(const char *, const char *, ssize_t, bool, char **);
+static int S3CacheExists(WT_FILE_SYSTEM *, const char *, bool *);
+static int S3CachePath(WT_FILE_SYSTEM *, const char *, char **);
+static int S3Path(const char *, const char *, char **);
+static int S3Exist(WT_FILE_SYSTEM *, WT_SESSION *, const char *, bool *);
 static int S3CustomizeFileSystem(
   WT_STORAGE_SOURCE *, WT_SESSION *, const char *, const char *, const char *, WT_FILE_SYSTEM **);
 static int S3AddReference(WT_STORAGE_SOURCE *);
 static int S3FileSystemTerminate(WT_FILE_SYSTEM *, WT_SESSION *);
 
+/*
+ * S3Exist --
+ *     Return if the file exists. First checks the cache, and then the S3 Bucket.
+ */
+static int
+S3Exist(WT_FILE_SYSTEM *fileSystem, WT_SESSION *session, const char *name, bool *existp)
+{
+    S3_STORAGE *s3;
+    int ret;
+    s3 = FS2S3(fileSystem);
+    S3_FILE_SYSTEM *s3Fs = (S3_FILE_SYSTEM *)fileSystem;
+    *existp = false;
+
+    /* First check to see if the file exists in the cache. */
+    if ((ret = S3CacheExists(fileSystem, name, existp)) != 0)
+        return (ret);
+
+    /* It's not in the cache, try the s3 bucket. */
+    if (!*existp)
+        ret = s3Fs->conn->ObjectExists(s3Fs->bucketName, name, *existp);
+
+    return (ret);
+}
+
+/*
+ * S3Path --
+ *     Construct a pathname from the directory and the object name.
+ */
+static int
+S3Path(const char *dir, const char *name, char **path)
+{
+    /* Skip over "./" and variations (".//", ".///./././//") at the beginning of the name. */
+    while (*name == '.') {
+        if (name[1] != '/')
+            break;
+        name += 2;
+        while (*name == '/')
+            name++;
+    }
+    int strSize = std::snprintf(nullptr, 0, "%s/%s", dir, name) + 1; // Extra space for '\0'
+    if (strSize <= 0)
+        return (1);
+
+    auto size = static_cast<size_t>(strSize);
+    char *buf = new char[size];
```

Comment:
let's use malloc/calloc - free for all non-class allocations for consistency. If new fails it throws an exception, we cant throw exceptions in this extension. It will be easier to check for failures in memory allocation with c styled allocation routines.